src/rgw/rgw_common.h

```diff
@@ -2423,4 +2441,37 @@ int decode_bl(bufferlist& bl, T& t)
   return 0;
 }
 
+static inline void start_trace(span_structure&& ss, Span&& span, req_state* const s, const char* name, bool should_store)
+{
+  #ifdef WITH_JAEGER
```

Comment:
put the ifdef around the function, then you will be sure it is not invoked if not compiler with the flag

Reply:
in some places I have code like this 
```
Span span;
start_span();
func_to_span();
finish_span();
```
I m not putting #ifdef here because then it will be very dirty, for example the above code will convert to this
```
#ifdef WITH_JAEGER
Span span;
start_trace();
funct_to_trace();
finish_trace();
#else
funct_to_trace();
#endif
```
the second code is just adding extra lines of code so I though I adding `#ifdef` in the `start_trace()` itself to keep my code clean

Reply:
but then you are unnecessarily create the span object and call the functions.
also, you will have to compile these functions, so you are not able to macro out tracing at cmake level

Reply:
okay I will update the method my code then

Reply:
@yuvalif I found the similar example with `LTTNG` they are not checking its directives before tracing for example
```
if (!index_op->is_prepared()) {
    tracepoint(rgw_rados, prepare_enter, req_id.c_str());
    r = index_op->prepare(CLS_RGW_OP_ADD, &state->write_tag, y);
    tracepoint(rgw_rados, prepare_exit, req_id.c_str());
    if (r < 0)
      return r;
  }
```
this is why I implemented it like this :)

Also one more question, if I dont define my function `start_trace` under `#ifdef` then will it make my code slower, as the compiler will compile this function and every time start_trace function is invoked it will called, but wont make traces when due to presence of `#ifdef` inside the function

Reply:
I don't know why it was done like that in LTTNG (maybe @cbodley know?)
the main reason that we want the ifdef to wrap around you entire code is because people are compiling are code for all sorts of targets, some of them may not support jaeger at all.
we don't want to block them if they don't require tracing.

Reply:
okay I understand now, I will implement the spans under `#ifdef`

Reply:
@alimaredia is working on packaging jaeger and its dependencies, with that jaeger will be installed in system along with ceph, so I think if we dont need to consider the case where jaeger is not installed.
Also if I implement my function like this `func(Span span=nullptr)` pasing span as optional paramter then for this to work completely jaeger must be installed in the system.

Reply:
noted

Reply:
from my experience with using kafka libraries in rgw, there are always some targets that won't support it, and we won't be able to package for (ubuntu16.04, freebsd, archlinux!?!)

Reply:
> in some places I have code like this
> 
> ```
> Span span;
> start_span();
> func_to_span();
> finish_span();
> ```
> 
> I m not putting #ifdef here because then it will be very dirty, for example the above code will convert to this
> 
> ```
> #ifdef WITH_JAEGER
> Span span;
> start_trace();
> funct_to_trace();
> finish_trace();
> #else
> funct_to_trace();
> #endif
> ```
> 
> the second code is just adding extra lines of code so I though I adding `#ifdef` in the `start_trace()` itself to keep my code clean

If we use [RAII](https://en.cppreference.com/w/cpp/language/raii) we can have a cleaner solution:
```cpp
#ifdef WITH_JAEGER
class SpanTracer {
  Span& span;
  SpanTracer(Span &_span) : span(_span) {
    span.start();
  }
  ~SpanTracer() {
    span.stop();
  }
  #define START_TRACE {SpanTracer __span_tracer;
  #define STOP_TRACE }
};
#else
  #define START_TRACE {
  #define STOP_TRACE }
#endif

...

START_TRACE
code_to_trace()
STOP_TRACE
```

Reply:
@yuvalif thanks for this :)

Reply:
@yuvalif one thing I m not understanding is, suppose I load function with optional paramter `Span`
`func(int, int, Span)` I m not putting any kind of directives here that means this code wont compile in system which doesnt have jaeger, to overcome this now we will have to do this 
```
if JAEGER
func(int int Span)
else
func(int int)
```
and similarly if we have to define the same two function in .cc file,  isnt this messy?

Reply:
okay I think this ca be done
```
struct optional_span{
if JAEGER
  Span
endd;
}
func(int int optional_span = nullptr)

Reply:
@yuvalif 
"#ifdef WITH_JAEGER
class SpanTracer {
  Span& span;
  SpanTracer(Span &_span) : span(_span) {
    span.start();
  }
  ~SpanTracer() {
    span.stop();
  }
  #define START_TRACE {SpanTracer __span_tracer;
  #define STOP_TRACE }
};
#else
  #define START_TRACE {
  #define STOP_TRACE }
#endif

...

START_TRACE
code_to_trace()
STOP_TRACE
"